text/0000-standard-dependency-metadata-format.md

```diff
@@ -0,0 +1,48 @@
+# Standardized Paketo Dependency Metadata Format
+
+## Summary
+
+This proposal suggests that the Paketo project should converge on a single dependency metadata format.
+
+## Motivation
+
+There are several projects, such as the introduction of ARM or the removal of stacks, that are being discussed that would force Paketo to update the metadata for dependencies. It seems prudent that if we are going to have to do a large update to dependency metadata that it might also be a good time for us to converge as a project on the dependency metadata format that we should be using going forward.
+
+This could also be a good first step in converging on shared tooling. By having a shared dependency metadata format we would have a good common convergence point to begin building universal tooling for the Paketo project. 
+
+## Implementation
+
+The following in the proposed metadata format:
+
+```toml
+[[metadata.dependencies]]
```

Comment:
This is the example I mentioned in terms of how we used dependencies in the ubi-nodejs-extension - https://github.com/paketo-community/ubi-nodejs-extension/blob/7caa04ee9a7e22dd1948c62fcaca03679be0db99/extension.toml#L15

```

  [[metadata.dependencies]]
    id = "node"
    name = "Ubi Node Extension"
    stacks = ["io.buildpacks.stacks.ubi8"]
    source = "paketocommunity/run-nodejs-18-ubi-base"
    version = "18.1000"
```

In the case of of the Node.js extension, dependencies are currently handled as follows:
- build - list of os packages to installed baked into extension
- run - pre-bundled into run image onto which application is 'switched'. In this case we use the info in the dependencies to decide which run image to use based on the Node.js Version requested.  This is similar to how the existing buildpacks use the metadata to find the Node.js package to use, except that we set the source to the name of a docker image instead of a tar.gz to install.

Reply:
The was we use the depdencies, it would be good if `checksum` was optional.

Reply:
I wonder if we could achieve the same behaviour for ubi extensions with something more explicit and less ad hoc... I.e. for "normal" dependencies it seems legit to require a checksum. I don't know, maybe make source and checksum one way, but offer another for the *virtual* (for lack of better word) dependencies that will be provided by the special run image?

Reply:
I second what @loewenstein is saying. This feels like a bit of a different use case. The metadata up for discussion here has always been associated with buildpacks that install dependencies directly. I'm sure there are assumptions built into that process and generalizing the metadata to support other installation mechanisms is going to complicate things or possibly create bugs due to age-old assumptions. While this seems like a small thing, I think it's also going to cause a lot of additional work to be required, because of past assumptions.

I 100% agree that this is a problem though. There's no way for a buildpack author to really know what's going to be in an image that was modified by an extension. I had hoped that upstream would provide some mechanism for this, but I'm not aware of one. In the absence of something upstream, I'm not opposed to creating our own solution (and possibly pushing to have that made the upstream solution). I'd want to talk more with upstream before we do anything though, to confirm they're not planning something, and also make that plan its own RFC.

Reply:
I agree it is different in some ways, but it is similar in the search for a match of the dependency, currently we can re-used the same supporting methods versus having to duplicate it. See below:

https://github.com/paketo-community/ubi-nodejs-extension/blob/7caa04ee9a7e22dd1948c62fcaca03679be0db99/generate.go#L62
```
				dependency, err := dependencyManager.Resolve(extensionFilePath, entry.Name, version, context.Stack)
```

Reply:
Maybe there could be the concept of different types of dependencies, each with their own required fields, and with a common set of required fields that are used by the Resolve?

Reply:
tl;dr - I think this all merits more discussion and I don't want to hold up this RFC on that discussion. It's too much scope creep IMHO.

Breaking it down more...

- You can make a case that the dependencies we install are similar to what's being installed through the extension. 
- You could also make a case they are different, specifically that what gets installed from a dependency is not the same as what gets installed via a package (sometimes one dependency is split into multiple packages, debug, source, docs, etc.). 
- There are also differences in the source of those dependencies, and do we need to make distinctions about that? 
- There's also the build plan, and can we use that more effectively to communicate buildpack needs?
- You could also make a case that what both buildpacks and extensions care about is certain tools or binaries that get installed, not the dependency or package itself (sometimes these are the same, but not always). i.e. I want `bash` or I want `javac` and I don't care what package gives me that. 

I think it's worth rethinking this process and what buildpacks need, rather than trying to inject this into the existing process.

Reply:
I think that I agree with the idea that we should maybe push this down the road and worry about this at a later date. I would like to see more mature implementation of this before committing to a defined schema.